pkg/volume/csi/csi_client.go

```diff
@@ -608,13 +610,19 @@ func (c *csiDriverClient) NodeGetVolumeStats(ctx context.Context, volID string,
 		VolumePath: targetPath,
 	}
 
+	var volumeCondition *csipbv1.VolumeCondition
```

Comment:
I feel a little uncomfortable by having a new field being returned. I am thinking if it actually makes sense to add the volumeCondition field into the volume.Metrics type(and rename it to Status?).
So that 

1. we dont have to change a whole bunch of signature here. Or maybe we would because we need to change the name?
2. It will help us in the future if we want to add more fields to this.

Reply:
First of all, I have the same feeling with you
> I feel a little uncomfortable by having a new field being returned

I also contacted with @xing-yang about this approach IMO,

1. The current approach is the simplest way to reach our original target and move the responsibility of the external-health-monitor agent to kubelet by code freeze of v1.21 beta. And it's also the way which takes the smallest change for current kubelet implementation
2. The real issue is not in design of GetMetrics or MetricsProvider interface. The issue is we choose to re-use the volumeCondition field which is already gotten in GetMetrics function. If we want to have a more clear way to get volumeCondition, we can add new interface in pkg/volume/volume.go called StateProvider that contains a method called GetVolumeStats. And then add an argument using *csipbv1.VolumeUsage type to GetMetrics. After that we can remove the calling of NodeGetVolumeStats function inside GetMetrics function, call it out of side and pass the volumeUsage by argument.

```go
// Volume represents a directory used by pods or hosts on a node. All method
// implementations of methods in the volume interface must be idempotent.
type Volume interface {
	// GetPath returns the path to which the volume should be mounted for the
	// pod.
	GetPath() string

	// MetricsProvider embeds methods for exposing metrics (e.g.
	// used, available space).
	MetricsProvider
        
        // StateProvider
        StateProvider
}
```

type StateProvider interface {
	GetState() (*Metrics, *csipbv1.NodeVolumeStatsResp, error)
}

func (s *volumeStatCalculator) calcAndStoreStats() {
	// Find all Volumes for the Pod
	volumes, found := s.statsProvider.ListVolumesForPod(s.pod.UID)
	if !found {
		return
	}

	// Get volume specs for the pod - key'd by volume name
	volumesSpec := make(map[string]v1.Volume)
	for _, v := range s.pod.Spec.Volumes {
		volumesSpec[v.Name] = v
	}

	// Call GetMetrics on each Volume and copy the result to a new VolumeStats.FsStats
	var ephemeralStats []stats.VolumeStats
	var persistentStats []stats.VolumeStats
	for name, v := range volumes {
                 csiVolumeStats :=  v.GetState()
		metric, csiVolumeStats, err := v.GetMetrics(csiVolumeStats.GetUsage())
                 ......

                	if utilfeature.DefaultFeatureGate.Enabled(features.CSIVolumeHealth) {
			if csiVolumeStats.VolumeCondition.Abnormal {
				s.eventRecorder.Event(s.pod, v1.EventTypeWarning, "VolumeConditionAbnormal", csiVolumeStats.VolumeCondition.GetMessage())
			}
		}
              .....

Reply:
I prefer to add a new interface to implement the GetVolumeState function, other than modify the Metrics struct type. volume.Metrics should only include the metrics info

Reply:
We are calling one RPC NodeGetVolumeStats to get both metrics and volume condition, so we need one function that returns both results and then process them separately.